### DISCUSSION TITLE
Concurrency and Virtual Threads

### DISCUSSION THREAD

**Discussion by jhg023** [spring-projects/spring-kafka#2915] [Q&A] [answered]
My service creates a `ConcurrentKafkaListenerContainerFactory` that's used by multiple methods marked with `@KafkaListener`.

I recently switched my service over to use virtual threads, and enabled them for Kafka as well:
```java
@Bean
public <V> ConcurrentKafkaListenerContainerFactory<String, V> kafkaListenerContainerFactory(
    ObjectMapper objectMapper
) {
    var factory = new ConcurrentKafkaListenerContainerFactory<String, V>();

    factory.setAutoStartup(true);
    factory.setConsumerFactory(kafkaConsumerFactory(objectMapper));

    factory.setContainerCustomizer(container -> {
        container.getContainerProperties().setAckMode(ContainerProperties.AckMode.MANUAL_IMMEDIATE);
        container.getContainerProperties().setListenerTaskExecutor(createVirtualThreadAsyncTaskExecutor());
    });

    return factory;
}

private AsyncTaskExecutor createVirtualThreadAsyncTaskExecutor() {
    var asyncTaskExecutor = new SimpleAsyncTaskExecutor();

    asyncTaskExecutor.setConcurrencyLimit(SimpleAsyncTaskExecutor.UNBOUNDED_CONCURRENCY);
    asyncTaskExecutor.setVirtualThreads(true);

    return asyncTaskExecutor;
}
```

With this pattern, should I be invoking `factory.setConcurrency`, or is that not needed anymore?

**Comment by garyrussell** [comment, marked as answer]
You still need to set concurrency if you want to consume multiple records from different partitions concurrently; each consumer will get its own virtual thread.

**Reply by jhg023** [reply]
Thanks, Gary!

**Comment by Liveitabhi14** [comment]
Is it necessary to set the `listenerTaskExecutor` by ourselves in the `ConcurrentKafkaListenerContainer` to use virtual threads?
Shouldn't the property `spring.threads.virtual.enabled=true` do it instead?

**Reply by sobychacko** [reply]
If you are using Spring Boot, you can use the property `spring.thread.virtual.enabled` and get a container auto-configured with virtual threads enabled. See this Spring Boot issue for more details: https://github.com/spring-projects/spring-boot/issues/36396. Behind the scenes, Boot does the same in the programmatic way shown above.
